[GFC][Integration] Update formatting context geometries before layout

sammygill opened the pull request:
#### eec4af42ce3ad9d112eb3f81b124f34e6191988f
<pre>
[GFC][Integration] Update formatting context geometries before layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=299806">https://bugs.webkit.org/show_bug.cgi?id=299806</a>
<a href="https://rdar.apple.com/161579865">rdar://161579865</a>

Reviewed by Alan Baradlay.

Adds some support for setting the geometry of relevant boxes in GFC. We
run this before we enter layout via updateFormattingContextGeometries.

Canonical link: <a href="https://commits.webkit.org/300862@main">https://commits.webkit.org/300862@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7e5121d85db845635a59beb102769728249c6ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34519 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130936 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76230 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/426adbf9-603a-42bb-bd04-647a61184c17) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125986 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52407 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94406 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127063 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111015 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74998 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f19a0bf7-a8d2-4fc1-bd6e-eef928fcc3cf) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29179 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74422 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105229 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29398 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133612 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51044 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38893 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102878 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51423 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107234 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102685 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26122 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48040 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26288 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47923 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50900 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56672 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50361 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53709 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52035 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/layout/integration/grid/LayoutIntegrationGridLayout.cpp:
```diff
@@ -45,6 +46,13 @@ GridLayout::GridLayout(RenderGrid& renderGrid)
+{
+    auto boxGeometryUpdater = BoxGeometryUpdater { layoutState(), gridBox() };
+    boxGeometryUpdater.setFormattingContextRootGeometry(CheckedRef { gridBoxRenderer() }->containingBlock()->contentBoxLogicalWidth());
+    boxGeometryUpdater.setFormattingContextContentGeometry(CheckedRef { layoutState() }->geometryForBox(gridBox()).contentBoxWidth(), { });
```
You should introduce `checkedGridBoxRenderer` and `checkedLayoutState` instead of wrapping in `CheckedRef`

alanbaradlay replied:
is this a repeating pattern?

nt1m replied:
That's generally the pattern we use across the codebase to access checked/protected members

alanbaradlay replied:
sorry, what I meant was if Sammy thinks if it was a one time thing or it repeats across the GFC codebase. I would not add it unless it is used everywhere (it's not a great pattern).

sammygill replied:
Not entirely sure how much, if any, it will continue to repeat. Happy to add it if we continue to use it! Also FWIW I see both usages throughout the codebase and it's not really clear (at least to me) when one should use one vs the other

nt1m replied:
See https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-passing-an-object-to-non-trivial-function-hold-a-smart-pointer-to-this-object-on-the-stack

checkedFoo() / protectedFoo() is the right pattern

sammygill replied:
Looks like I'll need it at least one more time so I'll add it before I land the patch

sammygill committed:
[GFC][Integration] Update formatting context geometries before layout
https://bugs.webkit.org/show_bug.cgi?id=299806
rdar://161579865

Reviewed by Alan Baradlay.

Adds some support for setting the geometry of relevant boxes in GFC. We
run this before we enter layout via updateFormattingContextGeometries.

Canonical link: https://commits.webkit.org/300862@main
